### DISCUSSION TITLE
AsyncSession does not throw an error when committed outside async context manager

### DISCUSSION THREAD

**Discussion by noahkw** [sqlalchemy/sqlalchemy#6528] [Usage Questions]
**Describe the bug**
When accidentally committing an AsyncSession outside an async context manager, no exception is thrown. This can lead to difficult to debug bugs.

**To Reproduce**

```py
import asyncio

from sqlalchemy import Column, Integer, String, select, delete
from sqlalchemy.orm import declarative_base, sessionmaker
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession


Base = declarative_base()


class Test(Base):
  __tablename__ = 'test'
  id = Column(Integer, primary_key=True, unique=True, autoincrement=True)
  value = Column(String)

engine = create_async_engine("sqlite+aiosqlite://")
Session = sessionmaker(bind=engine, class_=AsyncSession)


async def main():
  async with engine.begin() as conn:
    await conn.run_sync(Base.metadata.create_all)

  async with Session() as session1:
    session1.add(Test(value="foo"))
    await session1.commit()

  async with Session() as session2:
    statement = delete(Test).where(Test.value == "foo")
    await session2.execute(statement)
  
  # accidentally un-indented call to AsyncSession.commit()
  await session2.commit()

  async with Session() as session3:
    statement = select(Test).where(Test.value == "foo")
    (await session3.execute(statement)).one()

asyncio.run(main())
```

**Error**
An error should be thrown when calling `await session2.commit()`, because the session is closed and the changes are not actually committed to the database. Consequently, `(await session3.execute(statement)).one()` does not throw a `NoResultFound` exception, which is unexpected.

**Comment by CaselIT** [comment]
Hi,

I think this is working as expected and documented. The session will re-open itself after a `close()` is called if it's used again, so it's expected that the commit does not raises an error.

Regarding your example, the issue is not that the `commit` is not issued, is that when the context manager exits the session is closed and a rollback is issued in that case. You can see the db interaction by passing `echo=True` to the create engine function.

I agree that in this case example the current behaviour may not the most appropriate one. Maybe we could add a flag to prevent this reopen-after-close behaviour or change this in a future release (for async session only?). @zzzeek what's your opinion?

**Reply by zzzeek** [reply]
the session doesn't actually have a "closed" state right now, so if there's any bug here, it's that the name ".close" is a bad name.   ".close()" refers to the underyling database resources that may be used by the Session, those are closed, as well as the state of any ORM objects that might be present in the session, these are evicted.   if there's any feature missing here, it's that it does not have a "please make this Session unusable" kind of mode.     

the [docs for close](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#closing) should be pretty clear on what close does, despite its less than ideal name.   We could do with more "Seealso" links to this section linked from Session.close() and AsyncSession.close() docstring.

I also don't see anything special about async vs. sync here.

**Reply by zzzeek** [reply]
see fd3f98ee38b496916122d1a9a29b02d59ca671f9 for clarified linking and docs on all sides that should make it clear that .close() does not prevent the Session from being used again.

whether a new mode of operation that *does* do this is a separate discussion, the thing that may make this more relevant than it has been for the past 15 years is the new emphasis on context managers.    Since I wouldnt want to change any of the context managers, the proposal would be a new config option for sessionmaker "finalize_on_close" that would work in conjunction with a method such as Session.finalize() , or something, that flips an additional flag which prevents any further operations on the session.

**Reply by CaselIT** [reply]
Yes, that was my suggestion wrt the flag. But I'm not sure we need it (and with it yet another option for session and the context managers)

> I also don't see anything special about async vs. sync here.

yes, I agree that the behaviour should be the same for sync and async.
